compiler/sem/semtypes.nim

```diff
@@ -47,6 +47,33 @@ proc reportMeta(c: PContext; info: TLineInfo; t: PType) =
   else:
     checkMeta(c, info, t)
 
+proc liftTypeClass(c: PContext, typ: PType, prev: PType,
+                   info: TLineInfo): PType =
+  ## Lifts a type class type from "bare" symbols of built-in types. If there's
+  ## nothing to lift returns the `typ`. `prev` is the type to re-use -- it may
+  ## be `nil`.
+  result = typ
+
+  let isMagic = typ.sym != nil and typ.sym.magic != mNone
+
+  case result.kind
+  of tyOrdinal, tyRange, tySequence, tySet, tyArray, tyLent, tyOpenArray,
+     tyVarargs, tyUncheckedArray:
+    if isMagic:
+      # this is the "raw", uninstantiated magic type -> it's
+      # a built-in type class
+      result = newOrPrevType(tyBuiltInTypeClass, prev, c)
+      result.flags.incl tfCheckedForDestructor
+      # add the generic type, but don't propagate the flags
+      result.sons = @[typ]
+
+  of tySink:
+    if isMagic:
+      localReport(c.config, info, SemReport(kind: rsemSinkIsNotATypeClass))
+      result = newOrPrevType(tyError, prev, c)
+  else:
+    discard
```

Comment:
Could be turned inside-out, although it put's the focus more on the `isMagic` check:
```suggestion
  if isMagic:
    case result.kind
    of tyOrdinal, tyRange, tySequence, tySet, tyArray, tyLent, tyOpenArray,
       tyVarargs, tyUncheckedArray:
      # this is the "raw", uninstantiated magic type -> it's
      # a built-in type class
      result = newOrPrevType(tyBuiltInTypeClass, prev, c)
      result.flags.incl tfCheckedForDestructor
      # add the generic type, but don't propagate the flags
      result.sons = @[typ]

    of tySink:
      localReport(c.config, info, SemReport(kind: rsemSinkIsNotATypeClass))
      result = newOrPrevType(tyError, prev, c)
    else:
      discard
```

Reply:
just my two cents, feel free to ignore, being `isMagic` forward is a good thing as it seems to be the first fact that we need to establish in order to "do the right thing".